File: android/app/src/main/java/com/bannerbridge/BannerAd.java
package com.bannerbridge;

import android.content.Context;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.view.ViewGroup.LayoutParams;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.MobileAds;
import android.util.Log;
import android.view.Gravity;
import android.graphics.Color;
import android.app.Activity;
import android.view.View;
import java.util.Map;
import java.util.HashMap;

public class BannerAd extends FrameLayout {
    private static final String TAG = "BannerAd";
    private static final String TEST_BANNER_ID = "ca-app-pub-3940256099942544/6300978111";
    
    private AdView adView;
    private TextView helloWorldText;

    public BannerAd(Context context) {
        super(context);
        Log.d(TAG, "Initializing BannerAd");
        setupHelloWorldText();
        setupBannerAd();
    }

    private void setupHelloWorldText() {
        Log.d(TAG, "Setting up Hello World text");
        helloWorldText = new TextView(getContext());
        helloWorldText.setText("Hello Anchal");
        helloWorldText.setTextSize(24);
        helloWorldText.setTextColor(Color.BLACK);
        helloWorldText.setGravity(Gravity.CENTER);
        
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
            LayoutParams.MATCH_PARENT,
            LayoutParams.WRAP_CONTENT
        );
        params.gravity = Gravity.TOP | Gravity.CENTER_HORIZONTAL;
        helloWorldText.setLayoutParams(params);
        
        addView(helloWorldText);
    }

    private void setupBannerAd() {
        Log.d(TAG, "Setting up banner ad");
        
        MobileAds.initialize(getContext(), initializationStatus -> {
            initializationStatus.getAdapterStatusMap().forEach((adapter, status) -> 
                Log.d(TAG, "Adapter: " + adapter + " Status: " + status.getInitializationState())
            );
            createAndLoadAd();
        });
    }

    private void createAndLoadAd() {
        Log.d(TAG, "Creating new AdView");
        
        // Clean up existing AdView
        if (adView != null) {
            adView.destroy();
            adView.removeAllViews();
            removeView(adView);
            adView = null;
        }
        
        // Get current activity for better mediation performance
        Activity currentActivity = null;
        if (getContext() instanceof ReactContext) {
            currentActivity = ((ReactContext) getContext()).getCurrentActivity();
        }
        
        if (currentActivity == null) {
            Log.e(TAG, "No activity found");
            return;
        }
        
        // Create new AdView with activity
        adView = new AdView(currentActivity);
        adView.setAdUnitId(TEST_BANNER_ID);
        adView.setAdSize(AdSize.MEDIUM_RECTANGLE);
        
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
            LayoutParams.WRAP_CONTENT,
            LayoutParams.WRAP_CONTENT
        );
        params.gravity = Gravity.CENTER;
        adView.setLayoutParams(params);
        
        adView.setVisibility(View.VISIBLE);
        
        adView.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                super.onAdLoaded();
                Log.d(TAG, "Ad loaded successfully!");
                
                helloWorldText.setText("Ad Loaded Successfully!");
                
                AdSize adSize = adView.getAdSize();
                if (adSize != null) {
                    int width = adSize.getWidthInPixels(getContext());
                    int height = adSize.getHeightInPixels(getContext());
                    
                    measure(
                        MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
                        MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY)
                    );
                    
                    requestLayout();
                    invalidate();
                }
            }

            @Override
            public void onAdFailedToLoad(LoadAdError error) {
                Log.e(TAG, "Ad failed to load. Error: " + error.getMessage());
                helloWorldText.setText("Ad Failed to Load: " + error.getMessage());
            }

            @Override
            public void onAdOpened() {
                Log.d(TAG, "Ad opened");
            }
            
            @Override
            public void onAdClosed() {
                Log.d(TAG, "Ad closed");
            }
            
            @Override
            public void onAdImpression() {
                Log.d(TAG, "Ad impression");
            }
            
            @Override
            public void onAdClicked() {
                Log.d(TAG, "Ad clicked");
            }
        });

        // Add AdView to layout before loading
        addView(adView);
        
        // Load the ad
        AdRequest adRequest = new AdRequest.Builder().build();
        Log.d(TAG, "Requesting ad load");
        adView.loadAd(adRequest);
    }

    @Override
    public void requestLayout() {
        super.requestLayout();
        post(() -> {
            measure(
                MeasureSpec.makeMeasureSpec(getWidth(), MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(getHeight(), MeasureSpec.EXACTLY)
            );
            layout(getLeft(), getTop(), getRight(), getBottom());
        });
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        Log.d(TAG, "BannerAd detached from window");
        if (adView != null) {
            adView.destroy();
            adView = null;
        }
    }
}

class BannerAdManager extends SimpleViewManager<BannerAd> {
    public static final String REACT_CLASS = "BannerAd";
    
    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    public BannerAd createViewInstance(ThemedReactContext reactContext) {
        Log.d("BannerAdManager", "Creating new BannerAd instance");
        return new BannerAd(reactContext);
    }

    @ReactProp(name = "width")
    public void setWidth(BannerAd view, int width) {
        Log.d("BannerAdManager", "Setting width prop: " + width);
    }

    @ReactProp(name = "height")
    public void setHeight(BannerAd view, int height) {
        Log.d("BannerAdManager", "Setting height prop: " + height);
    }

    @Override
    public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
        Map<String, Object> builders = new HashMap<>();
        
        Map<String, String> onLoadEvent = new HashMap<>();
        onLoadEvent.put("registrationName", "onBannerAdLoaded");
        builders.put("topBannerAdLoaded", onLoadEvent);
        
        Map<String, String> onFailEvent = new HashMap<>();
        onFailEvent.put("registrationName", "onBannerAdFailedToLoad");
        builders.put("topBannerAdFailedToLoad", onFailEvent);
        
        return builders;
    }
} 

File: android/app/src/main/java/com/bannerbridge/BannerAd.kt
package com.bannerbridge

import android.content.Context
import android.widget.FrameLayout
import android.widget.TextView
import android.view.ViewGroup.LayoutParams
import com.facebook.react.bridge.ReactContext
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import android.util.Log
import android.view.Gravity
import android.graphics.Color
import android.util.DisplayMetrics
import android.widget.LinearLayout

class BannerAd(context: Context) : LinearLayout(context) {
    private var adView: AdView? = null
    private var helloWorldText: TextView? = null
    
    companion object {
        private const val TAG = "BannerAd"
        private const val TEST_BANNER_ID = "ca-app-pub-3940256099942544/6300978111"
    }
    
    init {
        orientation = VERTICAL
        Log.d(TAG, "Initializing BannerAd")
        setBackgroundColor(Color.WHITE)  // Change background back to white
        setupHelloWorldText()
        setupBannerAd()
    }

    private fun setupHelloWorldText() {
        Log.d(TAG, "Setting up Hello World text")
        helloWorldText = TextView(context).apply {
            text = "Hello Kotlin"
            textSize = 24f
            setTextColor(Color.BLACK)  // Change text color back to black
            setBackgroundColor(Color.TRANSPARENT)  // Make text background transparent
            gravity = Gravity.CENTER
            setPadding(20, 20, 20, 20)
            
            layoutParams = LinearLayout.LayoutParams(
                LayoutParams.MATCH_PARENT,
                LayoutParams.WRAP_CONTENT
            )
        }
        addView(helloWorldText)
        Log.d(TAG, "Hello World text added to view")
    }

    private fun setupBannerAd() {
        Log.d(TAG, "Setting up banner ad")
        
        MobileAds.initialize(context) { initializationStatus ->
            val statusMap = initializationStatus.adapterStatusMap
            statusMap.forEach { (adapter, status) ->
                Log.d(TAG, "Adapter: $adapter Status: ${status.initializationState}")
            }
            Log.d(TAG, "Mobile Ads SDK initialized, creating AdView")
            createAndLoadAd()
        }
    }

    private fun createAndLoadAd() {
        Log.d(TAG, "Creating new AdView")
        
        // Clean up existing AdView
        adView?.destroy()
        if (adView != null) {
            removeView(adView)
        }
        adView = null
        
        // Create new AdView
        adView = AdView(context).apply {
            adUnitId = TEST_BANNER_ID
            setAdSize(AdSize.MEDIUM_RECTANGLE)
            
            // Set layout params for LinearLayout to wrap content initially
            layoutParams = LinearLayout.LayoutParams(
                LayoutParams.WRAP_CONTENT,  // Let AdView determine its width based on AdSize
                LayoutParams.WRAP_CONTENT   // Let AdView determine its height based on AdSize
            ).apply {
                 gravity = Gravity.CENTER_HORIZONTAL // Center AdView horizontally within LinearLayout
            }
            // Don't set visibility here, will be added and made visible in onAdLoaded

            adListener = object : AdListener() {
                override fun onAdLoaded() {
                    super.onAdLoaded()
                    Log.d(TAG, "Ad loaded successfully!")
                    Log.d(TAG, "AdView size reported in onAdLoaded: ${width}x${height}")
                    Log.d(TAG, "AdSize reported in onAdLoaded: ${adSize?.width}x${adSize?.height}")
                    Log.d(TAG, "Parent size in onAdLoaded: ${this@BannerAd.width}x${this@BannerAd.height}")

                    // Update text when ad loads successfully
                    helloWorldText?.text = "Ad Loaded Successfully!"
                    
                    // Add AdView to layout after ad is loaded
                    this@BannerAd.addView(adView)

                    // Make AdView visible and request layout update
                    adView?.visibility = VISIBLE
                    this@BannerAd.requestLayout()
                }

                override fun onAdFailedToLoad(error: LoadAdError) {
                    super.onAdFailedToLoad(error)
                    Log.e(TAG, "Ad failed to load. Error code: ${error.code}")
                    Log.e(TAG, "Error message: ${error.message}")
                    Log.e(TAG, "Error domain: ${error.domain}")
                    // Optionally, update text or hide the container on failure
                    helloWorldText?.text = "Ad Failed to Load: ${error.getMessage()}"
                    // Hide AdView on failure
                    adView?.visibility = GONE
                }
                 // Add other potential AdListener methods for debugging
                 override fun onAdOpened() {
                     Log.d(TAG, "Ad opened")
                 }
                 override fun onAdClosed() {
                     Log.d(TAG, "Ad closed")
                 }
                 override fun onAdImpression() {
                     Log.d(TAG, "Ad impression")
                 }
                 override fun onAdClicked() {
                     Log.d(TAG, "Ad clicked")
                 }
            }
        }

        // Do NOT add AdView here, it will be added in onAdLoaded
        // addView(adView)
        
        // Load the ad
        val adRequest = AdRequest.Builder().build()
        Log.d(TAG, "Requesting ad load")
        adView?.loadAd(adRequest)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        Log.d(TAG, "BannerAd detached from window")
        adView?.destroy()
        adView = null
    }
}

class BannerAdManager : SimpleViewManager<BannerAd>() {
    override fun getName() = "BannerAd"

    override fun createViewInstance(reactContext: ThemedReactContext): BannerAd {
        Log.d("BannerAdManager", "Creating new BannerAd instance")
        val bannerAdView = BannerAd(reactContext)
        // Set initial layout params for the native view based on RN props
        val widthPx = reactContext.resources.displayMetrics.density * 300 // Convert dp to pixels
        val heightPx = reactContext.resources.displayMetrics.density * 250 // Convert dp to pixels
        bannerAdView.layoutParams = LinearLayout.LayoutParams(widthPx.toInt(), heightPx.toInt())
        return bannerAdView
    }

    @ReactProp(name = "width")
    fun setWidth(view: BannerAd, width: Int) {
        Log.d("BannerAdManager", "Setting width prop: $width")
        // React Native's layout system will handle the width based on the prop
    }

    @ReactProp(name = "height")
    fun setHeight(view: BannerAd, height: Int) {
        Log.d("BannerAdManager", "Setting height prop: $height")
        // React Native's layout system will handle the height based on the prop
    }
} 